### PR TITLE
List post duplicates (resolves #574). 

### LINKED ISSUE #574: ✨ Feature parity: list duplicates of a post
## How does this feature work on Reddit?

On both reddit.com and old.reddit.com, one can view discussions (== "duplicates") for a post on other subreddits by visiting either of these URLs:
* `/r/:sub/duplicates/:id`
* `/r/:sub/duplicates/:id/:title`
* `/duplicates/:id`
* `/duplicates/:id/:title`

where `:id` is the post ID, `:sub` the subreddit hosting the post, and `:title` the URL-friendly shortened title.

## Describe how this could be implemented into Libreddit
I've drafted a POC which I'll include below. The basic list of required modifications is:
* saving the value `num_duplicates` which is an integer in the post JSON object (== `[0].data.children[0]`)
* adding a link in the post to the duplicates that would point to `/r/:sub/duplicates/:id` but only if `num_duplicates > 0`
* creating a function that would `GET /r/:sub/duplicates/:id/.json` and retrieve and process the array of duplicate posts, which is (== `[1]data.children`)
* registering the appropriate API endpoints to call said function that would show a list of duplicates

[Here is my branch.](https://github.com/Daniel-Valentine/libreddit/tree/link-to-other-discussions) You can download it by doing:
```shell
git clone --branch=link-to-other-discussions -- https://github.com/Daniel-Valentine/libreddit
```

The page for a post will include a link for discussions (after the permalink) if duplicates exist for the post. That link points to `/r/:sub/duplicates/:id/:title` which will list those duplicates below the post. It's a little rough around the edges (it's my first foray into Rust and it's been a while since I've done Jinja2 templating), but I think it's adequate to show how it can be done. I can open up a PR into this repo's master if requested.

## Describe alternatives you've considered
Neither libreddit nor teddit offer this feature. The only alternative is to visit Reddit itself for a list of those duplicates, which would defeat the purpose of using a front-end like libreddit.

## Additional context / screenshot
Here is an example post with duplicates and a link to its duplicates:
* https://www.reddit.com/r/dataisbeautiful/comments/sm1zka/oc_most_popular_programming_languages_in_public/
* https://www.reddit.com/r/dataisbeautiful/duplicates/sm1zka/oc_most_popular_programming_languages_in_public/

These endpoints work on either reddit.com or old.reddit.com, and one can get the necessary JSON representation for either by adding `.json` to the end of each URL.

On the (new) web interface, the link to the duplicates appears above the comments block ("[View discussions in 3 other communities](https://www.reddit.com/r/dataisbeautiful/duplicates/sm1zka/oc_most_popular_programming_languages_in_public/)"). Following that link takes you to the post, but below will be a list of duplicates on other subreddits.

On the old web interface, the link to the same list is a tab to the right "comments" ("[other discussions (3)](https://old.reddit.com/r/dataisbeautiful/duplicates/sm1zka/oc_most_popular_programming_languages_in_public/)").

### PR DESCRIPTION
**Pull request by Daniel-Valentine**
This PR, a port of Ferrit PRs ferritreader/ferrit#22, implements the necessary changes to support listing a post's duplicates at `/r/:subreddit/duplicates/:id` and `/duplicates/:id`, thereby addressing libreddit/libreddit#574. To facilitate this:

* `src/duplicates.rs` is created, defining functions for handling the appropriate GET request and processing the JSON received from Reddit into a list of duplicate posts. parse_post has been moved to `src/utils.rs` since they're used by both `src/post.rs` and `src/duplicates.rs`.
* There is a new template `templates/duplicates.html` defining the page view for duplicates.
* `utils::filter_posts` has been modified to convey a number of posts filtered in addition to the bool informing whether or not all the posts were filtered. This allows for a message to let the user know that some duplicates were filtered out of the list.

Here is `/r/rust/duplicates/xv0xj7`:

Screenshot 2022-11-09 at 20-58-44 Initial Rust support is now merged into the Linux kernel! - r_rust

Like search results, the duplicates list is paginated (this is `/r/pics/duplicates/6z699p`):

Screenshot 2022-11-09 at 20-53-33 My dad waits every year for the day the sun rises just right and reflects along the railroad tracks Today was that day! - r_pics

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -511,15 +513,110 @@ pub fn get_filters(req: &Request<Body>) -> HashSet<String> {
...
+		score: format_num(score),
+		upvote_ratio: ratio as i64,
+		post_type,
+		media,
+		thumbnail: Media {
+			url: format_url(val(post, "thumbnail").as_str()),
+			alt_url: String::new(),
+			width: post["data"]["thumbnail_width"].as_i64().unwrap_or_default(),
+			height: post["data"]["thumbnail_height"].as_i64().unwrap_or_default(),
+			poster: "".to_string(),
```
Comment (sigaloid) [inline src/utils.rs:593]:
```suggestion
			poster: String::new(),
```

File: src/utils.rs [hunk 2]
```diff
@@ -511,15 +513,110 @@ pub fn get_filters(req: &Request<Body>) -> HashSet<String> {
...
+			background_color: val(post, "link_flair_background_color"),
+			foreground_color: if val(post, "link_flair_text_color") == "dark" {
+				"black".to_string()
+			} else {
+				"white".to_string()
+			},
+		},
+		flags: Flags {
+			nsfw: post["data"]["over_18"].as_bool().unwrap_or(false),
+			stickied: post["data"]["stickied"].as_bool().unwrap_or(false),
```
Comment (sigaloid) [inline src/utils.rs:611]:
```suggestion
			nsfw: post["data"]["over_18"].as_bool().unwrap_or_default(),
			stickied: post["data"]["stickied"].as_bool().unwrap_or_default(),
```

Comment (sigaloid) [reply src/utils.rs:611]: This is more correct but it's also used the other way in other places in the file, so feel free to reject

**Review by sigaloid — APPROVED** [review]

**Comment by Daniel-Valentine** [comment]
This PR now addresses ferritreader/ferrit#28 (porting ferritreader/ferrit#35).

Screenshot 2022-11-20 at 02-14-07 Initial Rust support is now merged into the Linux kernel! - r_rust

**Comment by sigaloid** [comment]
Awesome, looks great to be merged!

**Comment by Daniel-Valentine** [comment]
@spikecodes: What say you?

**Review by spikecodes — APPROVED** [review]

**Comment by Daniel-Valentine** [comment]
Seems to be a new bug with the navbar where it doesn't properly adjust to the size of the window. I'll fix it before this goes in.

**Pull request merged**